src/Features/CSharp/Portable/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -1,214 +1,101 @@
-                    syntaxNode = binaryExpression.Left.WalkDownParentheses();
-                }
-
-                yield return syntaxNode;
+                return SwitchExpressionArm(
+                    pattern: section.Labels.IsDefault
+                        ? DiscardPattern()
+                        : AsPatternSyntax(section.Labels[0].Pattern),
+                    whenClause: section.Labels.IsDefault
+                        ? null 
+                        : AsWhenClause(section.Labels[0]),
+                    expression: AsExpressionSyntax(section.Body));
             }
 
-            protected override IEnumerable<(ExpressionSyntax, StatementSyntax)> GetIfElseStatementChain(IfStatementSyntax currentStatement)
+            private static ExpressionSyntax AsExpressionSyntax(IOperation operation)
             {
-                StatementSyntax elseClauseStatement;
-                do
-                {
-                    yield return (currentStatement.Condition, currentStatement.Statement);
-                    elseClauseStatement = currentStatement.Else?.Statement;
-                    currentStatement = elseClauseStatement as IfStatementSyntax;
-                }
-                while (currentStatement != null);
-
-                if (elseClauseStatement != null)
+                return operation switch
                 {
-                    yield return (null, elseClauseStatement);
-                }
+                    IReturnOperation op => (ExpressionSyntax)op.ReturnedValue.Syntax,
+                    IThrowOperation op => ThrowExpression((ExpressionSyntax)op.Exception.Syntax),
+                    var v => throw ExceptionUtilities.UnexpectedValue(v.Kind)
+                };
             }
 
-            private static ExpressionSyntax GetLeftmostCondition(ExpressionSyntax syntaxNode)
+            public override SyntaxNode CreateSwitchStatement(SyntaxNode node, SyntaxNode expression, IEnumerable<SyntaxNode> sectionList)
             {
-                syntaxNode = syntaxNode.WalkDownParentheses();
-                while (syntaxNode.IsKind(SyntaxKind.LogicalAndExpression))
-                {
-                    syntaxNode = ((BinaryExpressionSyntax)syntaxNode).Left.WalkDownParentheses();
-                }
-
-                return syntaxNode;
+                var ifStatement = (IfStatementSyntax)node;
+                var block = ifStatement.Statement as BlockSyntax;
+                return SwitchStatement(
+                    switchKeyword: Token(SyntaxKind.SwitchKeyword).WithTriviaFrom(ifStatement.IfKeyword),
+                    openParenToken: ifStatement.OpenParenToken,
+                    expression: (ExpressionSyntax)expression,
+                    closeParenToken: ifStatement.CloseParenToken.WithPrependedLeadingTrivia(ElasticMarker),
+                    openBraceToken: block?.OpenBraceToken ?? Token(SyntaxKind.OpenBraceToken),
+                    sections: List(sectionList.OfType<SwitchSectionSyntax>()),
```

Comment:
1. when do we have a section list *not* built out of SwitchSectionSyntax?
2. if we have other types of nodes in that list, it looks like we're just droppin it on the floor.  is that ok?

Reply:
this is not changed in this PR. it should just use Cast (or generic if possible)

Reply:
i somewhat see thsi change as near a total rewrite.  So not a bad time to jsut clean these things up as well :)